Default arguments: implement captures for local functions, diagnose 'Self' for methods

slavapestov opened the pull request:
- Diagnose 'Self' in class method default arguments, since it's not supported
- Implement local function default argument captures
- Fix two other crashes found by inspection

I'll add a changelog entry in a new PR since I've already started tests on this one. This fixes stuff like

```
func outer(x: Int) {
  func inner(y: Int = x) {}

  inner() // equivalent to inner(y: x)
}
```

Fixes <https://bugs.swift.org/browse/SR-2189>, <rdar://problem/20796451>, <rdar://problem/55119566>.

slavapestov committed:
Sema: Compute and store captures for default argument expressions

slavapestov committed:
SIL: Refactor type lowering computation of the effective generic signature

This also fixes a long-standing bug with default argument expressions
capturing generic parameters.

slavapestov committed:
Sema: Diagnose captures of dynamic 'Self' type from default argument expressions

We could actually allow this for local functions, but it's not
worth implementing that until the more general ability of local
function default arguments to capture values is implemented.

Fixes <rdar://problem/55119566>.

slavapestov commented:
@swift-ci Please smoke test

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -5164,6 +5164,7 @@ class ParamDecl : public VarDecl {
     PointerUnion<Expr *, VarDecl *> DefaultArg;
     Initializer *InitContext = nullptr;
     StringRef StringRepresentation;
+    CaptureInfo Captures;
```
This is, uh, kind of expensive to put in every default argument, no?

slavapestov replied:
I guess, but I don't think there are that many default arguments. We can shrink it down to the size of a single ArrayRef with a bit of work though.

jrose-apple replied:
Maybe you'll have successfully nerd-sniped me into it after all!

slavapestov replied:
Hah, you're more than welcome to. I have another commit I'm going to tack onto this PR that implements default argument captures for local functions -- it wasn't too hard after all. Might want to wait for that one to avoid too many merge conflicts

slavapestov committed:
SILGen: Fix crash when immediately-applied closure transitively captures generic parameters

slavapestov committed:
SIL: Compute lowered local captures of SILDeclRefs and not AnyFunctionRefs

Unfortuantely this commit is bigger than I would like but I couldn't think
of any reasonable ways to split it up.

The general idea here is that capture computation is now done for a
SILDeclRef and not an AnyFunctionRef. This allows SIL to represent the
captures of a default argument generator.

slavapestov committed:
SILGen: Support for local function default argument captures

Fixes <https://bugs.swift.org/browse/SR-2189>, <rdar://problem/20796451>.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility
